### ISSUE TITLE
Explore data availability from EnhancAR paper

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/bolinas-dna#144]
[EnhancAR paper](https://www.biorxiv.org/content/10.64898/2026.04.13.718170v1) has a [Zenodo link](https://zenodo.org/records/19052620) to their processed data of Zoonomia cCRE homologs. We should check the feasibility of reformatting into our training dataset formats.

Some questions:
- Does it contain all cCREs? Does it contain metadata of cCRE of origin?
- Does it contain all species? Does it contain metadata of species?

**Comment by gonzalobenegas** [comment]
🤖 Inspected the [Zenodo bundle](https://zenodo.org/records/19052620). It is the **training corpus for the EnhancAR model** (Duncan et al. 2026, *"Evolutionary conditioning enables guided generation of functionally diverse enhancers"*, [bioRxiv 10.64898/2026.04.13.718170](https://www.biorxiv.org/content/10.64898/2026.04.13.718170v1)) — EnhancAR is the autoregressive design model; the bundle is what they fed it. Not a generic per-species enhancer release.

**Caveat:** my analysis is based on a streaming peek of the tarball + targeted spot-checks (one head-of-`train/` ~88 k-file index, one ~16 KB family, one ~16 KB family with broad species coverage). I did *not* fully extract or count val/test for resource-etiquette reasons. Numbers below quote the abstract / Zenodo metadata where stated; everything else is directly observed.

## What's in the bundle

`encode_screen_v4_enhancers.tar.gz`, **20.53 GB**, MD5 `8cdff2b6564831aa7aed511d58a93f5f`. Inside:

```
encode_screen_v4_proximal_and_distal_enhancers.no_resize/
├── train/   ← directly observed; partial streaming index = 88,929 .fa.aln files in train/ alone before I aborted
├── val/     ← inferred from the Zenodo description ("split into training, validation, and testing sets")
└── test/    ← inferred
```

No README, no manifest, no sidecar parquet/TSV — every member is `<EH38{D,F}_id>_<EH38E_id>.fa.aln`. Of 88,929 files in the partial index, **87,885 (98.8 %) used the `EH38D_*` form** (dataset accession) and **1,044 (1.2 %) used `EH38F_*`** (feature accession) — both pair with an `EH38E_*` (element) ID. The `.no_resize` suffix means cCREs were left at their native ENCODE coordinates rather than resized to a fixed window.

A typical record looks like (real example, family `EH38D3983568_EH38E4438019.fa.aln`):

```
>EH38D3983568_EH38E4438019::chr7:66016643-66016810  EH38D3983568_EH38E4438019::chr7:66016643-66016810  Homo_sapiens
aaccacagaaacctgtacagcagcatcagaggtctaccaatcacccaggagtgattgggtgattggagtctct...
>EH38D3983568_EH38E4438019::LT578344.1:60742257-60742425  EH38D3983568_EH38E4438019::LT578344.1:60742257-60742425  Gorilla_gorilla
aaccacagaaacctgtacagcagcatcagaggtctaccaatcacccaggagtgattgggtgattggagtctct...
>_R_EH38D3983568_EH38E4438019::CM001663.1:47990572-47990740  _R_EH38D3983568_EH38E4438019::CM001663.1:47990572-47990740  Nomascus_leucogenys
agccacagaaacctgtacagcagcatcagaagtctaccaaccacccaggagtgactgggtgattggagtctcg...
>_R_EH38D3983568_EH38E4438019::CM009269.2:15658335-15658499  _R_EH38D3983568_EH38E4438019::CM009269.2:15658335-15658499  Pongo_abelii
agccacagaaacctgtacagcagcatcagaggtctaccaatcacccaggagtgattggg----tggagtctcg...
```

Per record: cCRE accession (`EH38D…EH38E…`), contig (`chrom`/scaffold/INSDC), 0-based half-open coordinates (matches our convention), species (Zoonomia hub `Genus_species` style), and strand encoded by the `_R_` header prefix from `mafft --adjustdirection`. Sequences are gap-padded from the alignment; gaps strip cleanly with `s/-//g`.

## Direct answers to your four questions

> **Does it contain all cCREs?**

**Yes for ELS, no for the rest.** ENCODE SCREEN V4 GRCh38 has exactly 1,469,205 dELS + 249,464 pELS = **1,718,669 ELS**, and the abstract reports **"1.7 million human enhancer homolog sets"** — one set per ELS. Other cCRE classes (PLS, CTCF-only, CA-only, TF-only) are excluded by the dirname `proximal_and_distal_enhancers`. I did not see any `EH38E*` accession in train/ that wasn't an ELS, but I haven't enumerated all 1.7 M.

> **Does it contain metadata of cCRE of origin?**

**Yes, in the filename and in every record header.** The filename gives the `EH38E*` element ID (and the dataset/feature `EH38D|F` ID); each record's header repeats the accession plus the **human GRCh38 source coordinates** (the `Homo_sapiens` record carries `chr*:start-end` in UCSC chrom style). Trivial to round-trip back to the ENCODE SCREEN registry.

What's **not** in the bundle: cell-type labels. Every header in spot-checked records had exactly three whitespace-split fields (`<acc>::<contig>:<s>-<e>` + repeat + species) — no extra metadata. The "1,888 cell types" in the abstract is the cell-type *inventory* from SCREEN's per-cCRE annotations, not a per-record tag in this release. If we want cell types, we'd join back to ENCODE SCREEN by `EH38E*`.

> **Does it contain all species?**

**Species universe is the Zoonomia 241-way** (the [hub](http://cgl.gi.ucsc.edu/data/cactus/241-mammalian-2020v2-hub/) has exactly 241 genomes; species names in the FASTAs match its `Genus_species` convention). **Per-family coverage is variable** — only species where Cactus could project the human cCRE appear. One spot-checked family (`EH38D2725111_EH38E1718658`) had **40 species, all primates** — no rodents, carnivores, etc. — so coverage tracks evolutionary distance / regional conservation, not the full 241 in every family.

> **Does it contain metadata of species?**

Per record we get **species name** (`Genus_species`), **contig name**, **0-based half-open coordinates**, and **strand** (via `_R_`). Contig naming is mixed across species: some use Zoonomia hub-native scaffold IDs (`AloPal_scaffold_723`, `AteGeo_scaffold_6946`), some use INSDC accessions (`KZ207725.1`, `NTIC01038834.1`, `LT578344.1`, `CM001663.1`); the human record uses `chr*` UCSC names.

What's **not** stored explicitly: the **GenBank assembly accession** (e.g. `GCA_004027865.1`). The Zoonomia hub's `chromAlias.bb` (or the WGS-prefix → assembly lookup, e.g. `LLWD01` → *Acinonyx jubatus* `aciJub1`) closes this gap with a one-time mapping pass.

## Fit to [#121](https://github.com/Open-Athena/bolinas-dna/issues/121)

#121 proposed running `halLiftover` on the Zoonomia HAL ourselves (download ~1 TB HAL, `hal2fasta` per species, `halLiftover` from hg38 ENCODE SCREEN ELS to each Zoonomia target, package per-species BED + FASTA). This Zenodo bundle is essentially that output, pre-computed:

| #121 deliverable | Present? | Notes |
|---|---|---|
| Per-species coordinates on Zoonomia assembly | ✅ | In the FASTA headers — chrom/scaffold + 0-based half-open span. |
| Strand | ✅ | `_R_` header prefix. |
| Per-species sequence (`hal2fasta` equivalent) | ✅ | Ungap with `s/-//g` to recover. |
| Source human cCRE link | ✅ | ENCODE `EH38E*` accession in filename and every record header. |
| Source-set match (ENCODE V4 ELS) | ✅ | 1.7 M = exactly ENCODE V4 ELS count. |
| Species scope | ✅ ≥ #121's 20 | Up to 241; sparse per family. |
| Train/val/test split | ✅ | Theirs — confirm criterion before using (chromosome? family random?). Risk of leakage if any downstream eval splits differently. |
| Cell-type labels | ❌ | Source from ENCODE SCREEN by `EH38E*` accession if needed. |
| GenBank assembly accession per species | ❌ | One-time lookup via Zoonomia `chromAlias.bb` / WGS prefix. |
| Conservation/repeat masks | ❓ | Whatever Duncan et al. applied beyond ELS — check their methods. |

**Verdict: high feasibility, this can stand in for #121.** A reformatter would walk every `.fa.aln`, parse `(cCRE_id, contig, start, end, strand, species, sequence)` per record, gap-strip, and emit per-species parquet (or BED + FASTA), attaching assembly accession via a small lookup. Estimated effort: **S–M** — one library module + one Snakemake rule + a Zoonomia-name → GCA mapping table. We avoid the ~1 TB HAL download and the 240-species `halLiftover` compute that #121 was scoping.

## License caveat

Dataset is **CC BY-NC 4.0 (non-commercial)**. The paper itself is CC-BY but the dataset license is what binds use of these files. Worth confirming compatibility with our intended use before building on it.

## Suggested next step

If the license check passes, this looks like a clean substitute for #121, with the caveats above (assembly-accession mapping pass + accept their split or recompute). I'd suggest opening a follow-up issue for the reformatter and marking #121 as superseded once that lands. Happy to do the writeup of either.

---

*References used while writing this:* [Zenodo record API](https://zenodo.org/api/records/19052620) (file size, MD5, license); [bioRxiv API](https://api.biorxiv.org/details/biorxiv/10.64898/2026.04.13.718170/na/json) (abstract — Cloudflare blocks the HTML); [Zoonomia 241-way hub](http://cgl.gi.ucsc.edu/data/cactus/241-mammalian-2020v2-hub/) `genomes.txt` (241 genome confirmation) and one species' `chromAlias.txt` (assembly-mapping path); ENCODE V4 cCRE BED (`https://downloads.wenglab.org/Registry-V4/GRCh38-cCREs.bed`) for the ELS reference count.

**Comment by gonzalobenegas** [comment]
🤖 Two follow-ups, from one streaming `tar -xzO` of a single train/ family (`EH38D4202822_EH38E2701539.fa.aln`, 155,498 bytes — picked because it sits near the head of the tarball and is one of the larger ones I'd indexed):

## Repeat masking: **no** — case info has been destroyed

The sequences are **uniformly lowercase** — across the first 5,000 nucleotide-equivalent chars (covering ~6 species' records back-to-back), I found **5,000 lowercase, 0 uppercase, 0 N**. No `[A-Z]` anywhere in the file. The distinct characters in the sequence area (gaps stripped) are exactly `{a, c, g, n, t}`.

That means whoever built the bundle applied a uniform `tr A-Z a-z` somewhere in the pipeline, which **destroys** the standard soft-masking convention (lowercase = RepeatMasker repeat, uppercase = non-repeat) that `hal2fasta` would preserve from the Zoonomia 2bits. So we can't tell repeats from non-repeats from the FASTA alone — if we want repeat masks we'll have to re-apply them on the per-species coordinates after reformatting.

There **are** `N` runs in the data (max run length 17 in this sample) — those are presumably assembly gaps / hard-masked positions that survived because they're non-`[A-Z]` characters. Not many of them.

## Species scope: **not artificially capped** — varies with conservation

The 155 KB family has **202 records, 202 distinct species, no per-species duplication**. That's well above the 40-species (all-primate) family I'd looked at earlier — same release, same format, very different breadth. The species I see in this family span all the major Zoonomia clades (carnivores, cetartiodactyls, rodents, lagomorphs, afrotherians, xenarthrans, primates, marsupials/monotremes, …), which is consistent with no cap and just "include every species where Cactus could project the human cCRE".

So the answer is: **species count per family is set by alignability, not by a fixed cap.** Coverage range observed so far: 40 → 202 of the Zoonomia 241. A primate-only cCRE gets a primate-only family; a deeply conserved cCRE gets a near-241-wide family.

**Comment by gonzalobenegas** [comment]
🤖 Length stats across 7 spot-checked families (one streaming `tar -xz` of named members, no full extract). Sorted by file size:

```
EH38E suffix    kB recs  spp  dup  aln_w u_min u_p10 u_med u_p90 u_max  hum   cls encbed
E4438019         1    4    4    N    168   164   164   168   168   168  167  dELS    167
E4337021         2    6    6    N    363   344   344   345   361   361  345  dELS    345
E4171435         7   21   21    N    212   198   202   203   203   212  203  dELS    203
E3813382        20   39   39    N    402   154   222   322   325   325  322  dELS    322
E3028294        93  234  234    N    269   146   179   190   194   209  179  pELS    179
E2701539       151  202  202    N    630    45   281   294   316   354  316  dELS    316
E3828431       181  218  218    N    712   172   276   317   351   407  331  dELS    331
```

Columns: `recs` = FASTA records, `spp` = unique species, `dup` = any species appears >1×, `aln_w` = alignment column count, `u_*` = ungapped per-species length, `hum` = ungapped human record length, `cls`/`encbed` = ENCODE V4 cCRE class and BED length for cross-reference.

**Findings:**

- **Human ungapped length matches the ENCODE BED in 7/7** — sequences are at native ENCODE cCRE coordinates (the `.no_resize` dirname is honest). Human input length per family is whatever ENCODE assigned, bounded by ENCODE's hard `[150, 350]` bp window for SCREEN V4.
- **Per-species ungapped length clusters tightly around human length** — median ≈ human, p90 within ~5 % of human — with a **long left tail** of partial homologs going as low as 45 bp in one family. So Cactus sometimes projects only a fragment of the cCRE and the bundle keeps that fragment as-is, no re-padding.
- **No within-species duplication** in any of the 7 families — `n_records == n_species` everywhere. No paralogs / multi-projection records per genome.
- **Aligned width inflates with species diversity.** For a 4-species, near-human family `aln_w/hum` ≈ 1.01×; for the 234-species pELS it's 1.50×; for a broad 218-species dELS it's 2.15×. So if anyone uses the `.fa.aln` files directly, the *aligned* sequence length is family-specific and unbounded by the cCRE source size.
- **Both pELS and dELS represented** (1 pELS + 6 dELS in the sample), consistent with the dirname `proximal_and_distal_enhancers`.

**Bottom line on lengths:** for downstream training, the natural input is the **ungapped per-species sequence**, which is roughly `human_cCRE_length ± a few %` (so 150–350 bp, mostly 200–330 bp), with a thin tail of much-shorter partials. The alignment width on disk can be up to ~2× that — only relevant if we consume the alignment columns rather than per-species records.
